src/watchman_client.js

```diff
@@ -0,0 +1,376 @@
+
+  // define/clear some local state. The properties will be initialized
+  // in getClient(). This is also called again in _onEnd when
+  // trying to reestablish connection to watchman.Client.
+  this._clearLocalVars(); 
+
+  this._clientListeners = null;  // direct listeners from here to watchman.Client.
+
+  // we may want to call 'getClient()' and return a promise instead later,
+  // but for now we'll assume the WatchmanWatchers will not use wildmatch
+  // until they've done a getClient to retrieve the instance, so the
+  // capabilityCheck call is done.
+  Object.defineProperty(this, 'wildmatch', {
+    get() { return this._wildmatch; }
+  });
+}
+
+/**
+ * Set up the connection to the watchman client. If the optional
+ * value watchmanBinaryPath is passed in AND we're not already in the
+ * middle of setting up the client, use that as the path to the
+ * binary. In either case, return a promise that will be fulfilled
+ * when the client has been set up and capabilities have been returned.
+ */
+WatchmanClient.prototype.getClient = function(watchmanBinaryPath) {
+  
+  if (!this._clientPromise) {
+    this._clientPromise = new Promise((resolve, reject) => {
+
+      let client = new watchman.Client(watchmanBinaryPath
+                                       ? { watchmanBinaryPath: watchmanBinaryPath}
+                                       : {});
+
+      // reset the relevant local values
+
+      try {
+        client.capabilityCheck(
+          {optional: ['wildmatch', 'relative_root']},
+          (error, resp) => {
+            if (error) {
+              reject(error);
+            } else {
+              this._watchmanBinaryPath = watchmanBinaryPath;
+              this._wildmatch = resp.capabilities.wildmatch;
+              this._relative_root = resp.capabilities.relative_root;
+              this._client = client;
+
+              this._setupClientListeners();
+              resolve(this);
+            }
+          }
+        );
+      } catch (err) {
+        console.error("Client capabilityCheck threw error somehow: " + err);
+        // XXX Do something here to throw an error or reject?
+      }
+    });
+  }
+
+  return this._clientPromise;
+};
+
+/**
+ * Called from WatchmanWatcher (or this object during reconnect) to create
+ * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+ * watchman.Client, to be handled in this object.
+ */
+WatchmanClient.prototype.subscribe = function(watchmanWatcher, root) {
+
+  let watcherInfo = this._createWatcherInfo(watchmanWatcher);
+  let subscription = watcherInfo.subscription;
+  
+  return this._watch(subscription, root)
+    .then((resp) => this._clock(subscription))
+    .then((data) => this._subscribe(subscription))
+    .catch((error) => {
+      console.error("Caught error in subscribe for watchmanWatcher root " + root);
+      console.error(error);
+      return Promise.reject(error); // XXX is this right? Or do I just do 'reject(error)'?
```

Comment:
reject works, but you can also simply re-throw. `throw error;`

Im not sure we want the library to emit this logging by default, as the downstream consumer may correctly handle the error,  but have no way to hide the unwanted output.

Some options, logging can be done via node-debug instead, or the entire catch handler can be dropped. Deferring the problem to the caller of `subscribe`

Reply:
Thanks. I agree more with your 'let the caller decide what to do with the error, since we're not really doing anything interesting' idea. Removed the .catch() entirely.